### PR TITLE
User attribution fix

### PR DESCRIPTION
**Pull request by zutigrm**
## Summary

Addresses issue:

- #7892 

## Relevant technical choices

* There is a small deviation from IB, per this [comment](https://github.com/google/site-kit-wp/issues/7892#issuecomment-1833621117)
* php tests from #7922 were done as part of this issue

## PR Author Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 5.2 and PHP 5.6.
- [ ] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [ ] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [x] Run the code.
- [x] Ensure the acceptance criteria are satisfied.
- [x] Reassess the implementation with the IB.
- [x] Ensure no unrelated changes are included.
- [x] Ensure CI checks pass.
- [ ] Check Storybook where applicable.
- [x] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [ ] Ensure the PR has the correct target branch.
- [ ] Double-check that the PR is okay to be merged.
- [ ] Ensure the corresponding issue has a ZenHub release assigned.
- [ ] Add a changelog message to the issue.

### CODE REVIEW

File: tests/phpunit/integration/Core/User_Input/User_InputTest.php [hunk 1]
```diff
@@ -197,4 +207,64 @@ public function test_set_answers() {
...
+			array(
+				'purpose'       => array( 'other' ),
+				'postFrequency' => array( 'weekly' ),
+				'goals'         => array( 'improving_performance' ),
+			)
+		);
+
+		$existing_answers = $this->user_input->get_answers();
+		// Since purpose answer changed, it should be attributed to admin 2.
+		$this->assertEquals( $existing_answers['purpose']['answeredBy'], $second_admin_id );
```
Comment (tofumatt) [inline tests/phpunit/integration/Core/User_Input/User_InputTest.php:268]: Since we're changing all of the answers, let's check each one and make sure only the relevant ones changed attribution, just to be thorough 😄 

Basically the first question is the only site-wide one, so let's make sure the attribution for the other questions doesn't change here.

File: tests/phpunit/integration/Core/User_Input/User_InputTest.php [hunk 2]
```diff
@@ -197,4 +207,64 @@ public function test_set_answers() {
...
+			array(
+				'purpose'       => array( 'publish_blog' ),
+				'postFrequency' => array( 'weekly' ),
+				'goals'         => array( 'improving_performance' ),
+			)
+		);
+
+		$existing_answers = $this->user_input->get_answers();
+		// Since purpose answer didn't change, it should still be attributed to admin 1.
+		$this->assertEquals( $existing_answers['purpose']['answeredBy'], $this->user_id );
```
Comment (tofumatt) [inline tests/phpunit/integration/Core/User_Input/User_InputTest.php:238]: Maybe we should check that some changed here, and have one more test for what happens when _no_ answers change, just to cover all scenarios.

File: includes/Core/User_Input/User_Input.php [hunk 3]
```diff
@@ -228,8 +227,24 @@ public function set_answers( $settings ) {
...
+
+				if ( ! empty( $existing_answers['purpose']['values'] ) ) {
+					$answer_differs = array_diff( $existing_answers['purpose']['values'], $answers );
+					// If purpose answer changed, attribute it to the current user.
+					if ( ! empty( $answer_differs ) ) {
+						$answered_by = $this->user_options->get_user_id();
+					}
+				}
+
+				$setting_data['answeredBy']    = $answered_by;
```
Comment (tofumatt) [inline includes/Core/User_Input/User_Input.php:245]: I know the logic here makes sense, but I find it a bit tough to follow. In part because we assign `$answered_by` a default that should really only be the default when there's no answer given.

What do you think of this change? Or something like it? Just to follow the logic a bit more. Let me know 🙂 

```suggestion
				if (
					// If the answer to the "purpose" question changed,
					// attribute the answer to the current user changing the
					// answer.
					(
						! empty( $existing_answers['purpose']['values'] ) &&
						! empty( array_diff( $existing_answers['purpose']['values'], $answers ) )
					) ||
					// If the answer to the "purpose" question was empty,
					// attribute the answer to the current user.
					empty( $existing_answers['purpose']['answeredBy'] )
				) {
					$answered_by = $this->user_options->get_user_id();
				} else {
					// Otherwise, attribute the answer to the user who answered
					// the question previously.
					$answered_by = $existing_answers['purpose']['answeredBy'];
				}

				$setting_data['answeredBy'] = $answered_by;
```

**Review by tofumatt — CHANGES_REQUESTED** [review]
Looks good, just some thoughts on the tests and code here, but the logic seems right 👍🏻

File: includes/Core/User_Input/User_Input.php (hunk 3 above)
Comment (zutigrm) [reply includes/Core/User_Input/User_Input.php:245]: Thanks, this looks fine, with more explanations along the way, I will commit the change

File: tests/phpunit/integration/Core/User_Input/User_InputTest.php [hunk 4]
```diff
@@ -197,4 +207,75 @@ public function test_set_answers() {
...
+		$this->user_input->set_answers( $answers );
+
+		$existing_answers = $this->user_input->get_answers();
+		// Since purpose answer didn't change, it should still be attributed to admin 1.
+		$this->assertEquals( $existing_answers['purpose']['answeredBy'], $this->user_id );
+
+		$this->user_input->set_answers( $answers );
+		$existing_answers = $this->user_input->get_answers();
+		// Since no answer changed, purposr answer should still be attributed to admin 1.
+		$this->assertEquals( $existing_answers['purpose']['answeredBy'], $this->user_id );
```
Comment (tofumatt) [inline tests/phpunit/integration/Core/User_Input/User_InputTest.php:242]:
```suggestion
		$this->user_input->set_answers( $answers );
		$existing_answers = $this->user_input->get_answers();
		// Since no answer changed, the "purpose" answer should still be attributed to admin 1.
		$this->assertEquals( $existing_answers['purpose']['answeredBy'], $this->user_id );
```

Let's move this to its own test spec/function, just for clarity. I know it's technically testing the behaviour here, but it's a bit hard to follow the logic of this here. It's probably easier if we just have a new test spec that sets the same answers separate to this one 🙂

**Review by tofumatt — CHANGES_REQUESTED** [review]
Looks good! 👍🏻 

Just one minor comment then we can merge 🙂

File: tests/phpunit/integration/Core/User_Input/User_InputTest.php (hunk 4 above)
Comment (zutigrm) [reply tests/phpunit/integration/Core/User_Input/User_InputTest.php]: Makes sense, thanks.

PR Updated

**Review by tofumatt — APPROVED** [review]
Awesome, looks great 👍🏻
